File: impl/ruby/lib/naalp/cose.rb
# Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
#
# N-AALP C2 signing layer for the Ruby SDK: the COSE_Sign1 (RFC 9052) signing-input and
# object assembly, plus deterministic ML-DSA (FIPS 204, rnd=0) and Ed25519 (RFC 8032).
#
# The deterministic ML-DSA path uses the platform OpenSSL (>= 3.5) ML-DSA provider with the
# signature parameter deterministic=1 (which fixes the FIPS 204 rnd to 32 zero bytes) and a
# seed-only PKCS#8 private key (the [0] seed CHOICE), producing signatures and public keys
# byte-identical to the Go (CIRCL), Rust (fips204) and Python (dilithium-py) references —
# cross-checked against the shared conformance corpus and the NIST ACVP keyGen vectors.
require 'openssl'
require_relative 'cbor'

module Naalp
  module COSE
    include Naalp::CBOR

    ALG_MLDSA65 = -49
    ALG_MLDSA87 = -50
    ALG_ED25519 = -19

    PROFILE_PUBLIC = 1
    PROFILE_ENTERPRISE = 2
    PROFILE_SOVEREIGN = 3

    TAG_SIGN1 = 18

    # DER of the AlgorithmIdentifier SEQUENCE { OID } per parameter set (no params).
    MLDSA_OID_DER = {
      "ML-DSA-65" => ["300b0609608648016503040312"].pack("H*"),
      "ML-DSA-87" => ["300b0609608648016503040313"].pack("H*"),
    }.freeze
    ALG_TO_PARAM = { ALG_MLDSA65 => "ML-DSA-65", ALG_MLDSA87 => "ML-DSA-87" }.freeze

    module_function

    # NIST security level of a registered alg, and whether it is registered. Ed25519 is
    # classical (level 0), valid only as a hybrid leg. Returns [level, known].
    def alg_level(alg)
      case alg
      when ALG_MLDSA87 then [5, true]
      when ALG_MLDSA65 then [3, true]
      when ALG_ED25519 then [0, true]
      else [0, false]
      end
    end

    # Minimum signature level a profile accepts (Sovereign floors at level 5; else 3).
    def profile_min_level(profile)
      profile == PROFILE_SOVEREIGN ? 5 : 3
    end

    def to_be_signed_raw(protected_bytes, payload)
      Naalp::CBOR.encode(
        Naalp::CBOR::A.new([
          Naalp::CBOR::T.new("Signature1"),
          Naalp::CBOR::B.new(protected_bytes),
          Naalp::CBOR::B.new("".b),
          Naalp::CBOR::B.new(payload),
        ])
      )
    end

    def assemble_sign1_raw(protected_bytes, payload, sig)
      Naalp::CBOR.encode(
        Naalp::CBOR::Tag.new(TAG_SIGN1, Naalp::CBOR::A.new([
          Naalp::CBOR::B.new(protected_bytes),
          Naalp::CBOR::M.new([]),
          Naalp::CBOR::B.new(payload),
          Naalp::CBOR::B.new(sig),
        ]))
      )
    end

    def parse_sign1_raw(obj)
      v = Naalp::CBOR.decode(obj)
      unless v.is_a?(Naalp::CBOR::Tag) && v.n == TAG_SIGN1 && v.content.is_a?(Naalp::CBOR::A)
        raise "not a tagged COSE_Sign1"
      end
      arr = v.content.items
      unless arr.length == 4 && arr[0].is_a?(Naalp::CBOR::B) &&
             arr[2].is_a?(Naalp::CBOR::B) && arr[3].is_a?(Naalp::CBOR::B)
        raise "malformed COSE_Sign1 array"
      end
      [arr[0].v, arr[2].v, arr[3].v]
    end

    # --- COSE_Sign (tag 98) multi-signature support: the §5.2 Rotation object co-signature ---

    TAG_SIGN = 98

    # One COSE_Signature protected header: {1: alg} (RFC 9052 §4).
    def leg_protected(alg)
      Naalp::CBOR.encode(Naalp::CBOR::M.new([[Naalp::CBOR::U.new(1), Naalp::CBOR::N.new(alg)]]))
    end

    # The per-signer COSE_Signature signing input for a COSE_Sign (RFC 9052 §4.4):
    # det-CBOR(["Signature", body_protected, sign_protected, external_aad(empty), payload]). Note the
    # five-element "Signature" structure (with the per-leg sign_protected) vs the four-element
    # "Signature1" of a COSE_Sign1.
    def signature_to_be_signed(body_prot, signer_alg, payload)
      Naalp::CBOR.encode(Naalp::CBOR::A.new([
        Naalp::CBOR::T.new("Signature"),
        Naalp::CBOR::B.new(body_prot),
        Naalp::CBOR::B.new(leg_protected(signer_alg)),
        Naalp::CBOR::B.new("".b),
        Naalp::CBOR::B.new(payload),
      ]))
    end

    # Build one COSE_Signature leg: [leg_protected_bytes, signature_bytes].
    def signature_leg(body_prot, alg, seed, payload)
      sprot = leg_protected(alg)
      sig = mldsa_sign(alg, seed, signature_to_be_signed(body_prot, alg, payload))
      [sprot, sig]
    end

    # The tagged COSE_Sign object: 98([body_prot, {}, payload, [[sprot, {}, sig], ...]]).
    def assemble_sign_raw(body_prot, payload, legs)
      sig_arr = Naalp::CBOR::A.new(legs.map do |(sprot, sig)|
        Naalp::CBOR::A.new([Naalp::CBOR::B.new(sprot), Naalp::CBOR::M.new([]), Naalp::CBOR::B.new(sig)])
      end)
      Naalp::CBOR.encode(Naalp::CBOR::Tag.new(TAG_SIGN, Naalp::CBOR::A.new([
        Naalp::CBOR::B.new(body_prot), Naalp::CBOR::M.new([]), Naalp::CBOR::B.new(payload), sig_arr,
      ])))
    end

    # Recover [body_prot, payload, [[sprot, sig], ...]] from a tagged COSE_Sign object.
    def parse_sign_raw(obj)
      v = Naalp::CBOR.decode(obj)
      unless v.is_a?(Naalp::CBOR::Tag) && v.n == TAG_SIGN && v.content.is_a?(Naalp::CBOR::A)
        raise "not a tagged COSE_Sign"
      end
      arr = v.content.items
      unless arr.length == 4 && arr[0].is_a?(Naalp::CBOR::B) &&
             arr[2].is_a?(Naalp::CBOR::B) && arr[3].is_a?(Naalp::CBOR::A)
        raise "malformed COSE_Sign array"
      end
      legs = arr[3].items.map do |e|
        unless e.is_a?(Naalp::CBOR::A) && e.items.length == 3 &&
               e.items[0].is_a?(Naalp::CBOR::B) && e.items[2].is_a?(Naalp::CBOR::B)
          raise "malformed COSE_Signature leg"
        end
        [e.items[0].v, e.items[2].v]
      end
      [arr[0].v, arr[2].v, legs]
    end

    # Extract the alg (label 1) value from a serialized leg protected header {1: alg}.
    def alg_from_protected(prot)
      # §3.1.1 (R5): reject the redundant 0x41A0 encoding of an empty protected header (a bstr
      # wrapping an empty map; its unwrapped content is the single byte 0xA0) as NonCanonical,
      # before interpreting the header -- the empty protected header is pinned to 0x40.
      if prot.bytesize == 1 && prot.getbyte(0) == 0xA0
        raise Naalp::CBOR::NonCanonical, "empty protected header must be 0x40, not 0x41A0 (§3.1.1, R5)"
      end
      v = Naalp::CBOR.decode(prot)
      raise "protected header not a map" unless v.is_a?(Naalp::CBOR::M)
      v.pairs.each do |k, val|
        return val.v if k.is_a?(Naalp::CBOR::U) && k.v == 1 && val.is_a?(Naalp::CBOR::N)
      end
      raise "no alg in protected header"
    end

    # --- ML-DSA (FIPS 204) via OpenSSL ---

    # Build a seed-only PKCS#8 (OneAsymmetricKey) DER for an ML-DSA private key: the privateKey
    # OCTET STRING wraps the [0] IMPLICIT seed CHOICE (0x80 0x20 || 32-byte seed).
    def mldsa_seed_der(param, seed)
      raise "ml-dsa seed must be 32 bytes" if seed.bytesize != 32
      algid = MLDSA_OID_DER[param] or raise "unknown ML-DSA parameter set #{param}"
      seed_choice = [0x80, 0x20].pack("C*") + seed
      priv_os = [0x04, seed_choice.bytesize].pack("C*") + seed_choice
      inner = [0x02, 0x01, 0x00].pack("C*") + algid + priv_os
      [0x30, inner.bytesize].pack("C*") + inner
    end

    def mldsa_key_from_seed(param, seed)
      OpenSSL::PKey.read(mldsa_seed_der(param, seed))
    end

    def mldsa_keygen(param, seed)
      mldsa_key_from_seed(param, seed).raw_public_key
    end

    def mldsa_sign(alg, seed, tbs)
      param = ALG_TO_PARAM[alg] or raise "alg #{alg} is not an ML-DSA algorithm"
      key = mldsa_key_from_seed(param, seed)
      key.sign(nil, tbs, { "deterministic" => 1 })
    end

    def mldsa_verify(alg, pk, tbs, sig)
      param = ALG_TO_PARAM[alg] or raise "alg #{alg} is not an ML-DSA algorithm"
      pub = OpenSSL::PKey.new_raw_public_key(param, pk)
      pub.verify(nil, sig, tbs)
    end

    # --- Ed25519 (RFC 8032) via OpenSSL ---

    def ed25519_sign(seed, msg)
      raise "ed25519 secret key must be a 32-byte seed" if seed.bytesize != 32
      OpenSSL::PKey.new_raw_private_key("ED25519", seed).sign(nil, msg)
    end

    def ed25519_verify(pk, msg, sig)
      OpenSSL::PKey.new_raw_public_key("ED25519", pk).verify(nil, sig, msg)
    rescue OpenSSL::PKey::PKeyError
      false
    end

    # --- LAMPS opt-in composite signature (alg -65537, design.md §4.2) ---

    ALG_COMPOSITE_65_ED25519 = -65537 # COMPSIG-MLDSA65-Ed25519-SHA512
    ALG_COMPOSITE_44_ED25519 = -65538 # edge; RESERVED, not implemented
    COMPOSITE_PREFIX = "CompositeAlgorithmSignatures2025".b
    COMPOSITE_LABEL_MLDSA65_ED25519 = "COMPSIG-MLDSA65-Ed25519-SHA512".b
    MLDSA65_SIG_SIZE = 3309 # FIPS 204 ML-DSA-65 signature size
    MLDSA65_PUB_SIZE = 1952 # FIPS 204 ML-DSA-65 pubkey size (composite split point)

    # The LAMPS composite message representative M' = Prefix || Label || len(ctx) || ctx ||
    # SHA-512(M) (design.md §4.2). len(ctx) is a single length octet; the N-AALP composite context
    # is empty, so the octet is 0x00. Both legs sign this same M'.
    def compute_mprime(label, ctx, m)
      raise "composite context exceeds one length octet" if ctx.bytesize > 255

      COMPOSITE_PREFIX + label + [ctx.bytesize].pack("C") + ctx + OpenSSL::Digest::SHA512.digest(m)
    end

    # The LAMPS composite signature value over the COSE ToBeSigned tbs: mldsaSig || tradSig
    # (ML-DSA-65 first, raw concatenation; §4.2). The ML-DSA leg is deterministic (rnd=0) with
    # context = the suite Label octets (OpenSSL context-string); the Ed25519 leg signs M' with no
    # context.
    def composite_sign(mldsa_seed, ed_seed, tbs)
      mprime = compute_mprime(COMPOSITE_LABEL_MLDSA65_ED25519, "".b, tbs)
      key = mldsa_key_from_seed("ML-DSA-65", mldsa_seed)
      mldsa_sig = key.sign(nil, mprime,
                           { "deterministic" => 1, "context-string" => COMPOSITE_LABEL_MLDSA65_ED25519 })
      trad_sig = ed25519_sign(ed_seed, mprime)
      mldsa_sig + trad_sig # ML-DSA first (LAMPS order)
    end

    # Valid IFF BOTH the ML-DSA-65 leg (context = Label) and the Ed25519 leg (no context) validate
    # over M'. A value of the wrong length is malformed and rejected. A stripped or re-interpreted
    # lone leg has no valid composite because M' binds both components (RFC 9955; §4.2/§4.5).
    def composite_verify(mldsa_pk, ed_pk, m, sig)
      return false if sig.bytesize != MLDSA65_SIG_SIZE + 64

      mprime = compute_mprime(COMPOSITE_LABEL_MLDSA65_ED25519, "".b, m)
      pub = OpenSSL::PKey.new_raw_public_key("ML-DSA-65", mldsa_pk)
      mldsa_ok = pub.verify(nil, sig[0, MLDSA65_SIG_SIZE], mprime,
                            { "context-string" => COMPOSITE_LABEL_MLDSA65_ED25519 })
      ed_ok = ed25519_verify(ed_pk, mprime, sig[MLDSA65_SIG_SIZE..])
      mldsa_ok && ed_ok
    rescue OpenSSL::PKey::PKeyError
      false
    end

    def cose_sign1(alg, seed, protected_bytes, payload)
      tbs = to_be_signed_raw(protected_bytes, payload)
      sig = mldsa_sign(alg, seed, tbs)
      assemble_sign1_raw(protected_bytes, payload, sig)
    end

    # Verify a raw signature over already-assembled ToBeSigned bytes, dispatching by alg.
    def cose_verify1_raw(alg, pk, tbs, sig)
      case alg
      when ALG_MLDSA65, ALG_MLDSA87
        mldsa_verify(alg, pk, tbs, sig)
      when ALG_ED25519
        ed25519_verify(pk, tbs, sig)
      else
        raise "unknown alg #{alg}"
      end
    end

    def cose_verify1(alg, pk, obj)
      protected_bytes, payload, sig = parse_sign1_raw(obj)
      tbs = to_be_signed_raw(protected_bytes, payload)
      cose_verify1_raw(alg, pk, tbs, sig)
    end
  end
end

File: impl/kotlin/src/main/kotlin/sh/bubblefish/naalp/Cose.kt
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
package sh.bubblefish.naalp

import java.security.MessageDigest
import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters
import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters
import org.bouncycastle.crypto.params.MLDSAParameters
import org.bouncycastle.crypto.params.MLDSAPrivateKeyParameters
import org.bouncycastle.crypto.params.MLDSAPublicKeyParameters
import org.bouncycastle.crypto.params.ParametersWithContext
import org.bouncycastle.crypto.signers.Ed25519Signer
import org.bouncycastle.crypto.signers.MLDSASigner

/**
 * N-AALP C2 signing layer for the Kotlin SDK: the COSE_Sign1 (RFC 9052) signing-input and object
 * assembly, plus deterministic ML-DSA (FIPS 204, rnd=0) and Ed25519 (RFC 8032).
 *
 * The deterministic ML-DSA path uses Bouncy Castle's [MLDSASigner] initialised WITHOUT a
 * ParametersWithRandom, so the FIPS 204 `rnd` stays 32 zero bytes — byte-identical to the Go
 * (CIRCL), Rust (fips204), Python (dilithium-py) and Java (Bouncy Castle) reference implementations.
 * Key material is derived from the 32-byte NIST seed (xi) via the seed-only private-key constructor,
 * so the public key equals the NIST ACVP keyGen vector.
 */
object Cose {
    const val ALG_MLDSA65 = -49
    const val ALG_MLDSA87 = -50
    const val ALG_ED25519 = -19

    const val PROFILE_PUBLIC = 1L
    const val PROFILE_ENTERPRISE = 2L
    const val PROFILE_SOVEREIGN = 3L

    const val TAG_SIGN1 = 18L

    /**
     * NIST security level of a registered alg, paired with whether it is registered. ML-DSA-87 is
     * level 5, ML-DSA-65 is level 3, Ed25519 is classical (level 0, valid only as a hybrid leg);
     * any other alg is unregistered. Mirrors the Go/Rust/Python C2 registry.
     */
    fun algLevel(alg: Int): kotlin.Pair<Int, Boolean> = when (alg) {
        ALG_MLDSA87 -> kotlin.Pair(5, true)
        ALG_MLDSA65 -> kotlin.Pair(3, true)
        ALG_ED25519 -> kotlin.Pair(0, true)
        else -> kotlin.Pair(0, false)
    }

    /** Minimum signature level a profile accepts (Sovereign floors at level 5; else level 3). */
    fun profileMinLevel(profile: Long): Int = if (profile == PROFILE_SOVEREIGN) 5 else 3

    /** The RFC 9052 §4.4 Sig_structure for a COSE_Sign1 over an already-serialized header. */
    fun toBeSignedRaw(protectedHeader: ByteArray, payload: ByteArray): ByteArray =
        Cbor.encode(
            Cbor.A(
                listOf(
                    Cbor.T("Signature1"),
                    Cbor.B(protectedHeader),
                    Cbor.B(ByteArray(0)),
                    Cbor.B(payload)
                )
            )
        )

    /** The tagged COSE_Sign1 object: 18([protected, {}, payload, signature]). */
    fun assembleSign1Raw(protectedHeader: ByteArray, payload: ByteArray, sig: ByteArray): ByteArray =
        Cbor.encode(
            Cbor.Tag(
                TAG_SIGN1,
                Cbor.A(
                    listOf(
                        Cbor.B(protectedHeader),
                        Cbor.M(listOf()),
                        Cbor.B(payload),
                        Cbor.B(sig)
                    )
                )
            )
        )

    /** Recover [protected, payload, sig] from a tagged COSE_Sign1 object. */
    fun parseSign1Raw(obj: ByteArray): Array<ByteArray> {
        val v = Cbor.decode(obj)
        if (v !is Cbor.Tag || v.n != TAG_SIGN1 || v.content !is Cbor.A) {
            throw NaalpException("Malformed", "not a tagged COSE_Sign1")
        }
        val items = v.content.items
        if (items.size != 4 || items[0] !is Cbor.B || items[2] !is Cbor.B || items[3] !is Cbor.B) {
            throw NaalpException("Malformed", "malformed COSE_Sign1 array")
        }
        return arrayOf((items[0] as Cbor.B).v, (items[2] as Cbor.B).v, (items[3] as Cbor.B).v)
    }

    // --- COSE_Sign (tag 98) multi-signature support: the §5.2 Rotation object co-signature ---

    const val TAG_SIGN = 98L

    /** A parsed tag-98 COSE_Sign: body protected header, payload, and ordered [sprot, sig] legs. */
    class RotationParse(val bodyProt: ByteArray, val payload: ByteArray, val legs: List<Array<ByteArray>>)

    /** One COSE_Signature protected header: {1: alg} (RFC 9052 §4). */
    fun legProtected(alg: Int): ByteArray =
        Cbor.encode(Cbor.M(listOf(Cbor.Pair(Cbor.U(1), Cbor.N(alg.toLong())))))

    /**
     * The per-signer COSE_Signature signing input for a COSE_Sign (RFC 9052 §4.4):
     * det-CBOR(["Signature", body_protected, sign_protected, external_aad(empty), payload]). Note the
     * five-element "Signature" structure (with the per-leg sign_protected) vs the four-element
     * "Signature1" of a COSE_Sign1.
     */
    fun signatureToBeSigned(bodyProt: ByteArray, signerAlg: Int, payload: ByteArray): ByteArray =
        Cbor.encode(
            Cbor.A(
                listOf(
                    Cbor.T("Signature"),
                    Cbor.B(bodyProt),
                    Cbor.B(legProtected(signerAlg)),
                    Cbor.B(ByteArray(0)),
                    Cbor.B(payload)
                )
            )
        )

    /** Build one COSE_Signature leg: [leg_protected_bytes, signature_bytes]. */
    fun signatureLeg(bodyProt: ByteArray, alg: Int, seed: ByteArray, payload: ByteArray): Array<ByteArray> {
        val sprot = legProtected(alg)
        val sig = mldsaSign(alg, seed, signatureToBeSigned(bodyProt, alg, payload))
        return arrayOf(sprot, sig)
    }

    /** The tagged COSE_Sign object: 98([body_prot, {}, payload, [[sprot, {}, sig], ...]]). */
    fun assembleSignRaw(bodyProt: ByteArray, payload: ByteArray, legs: List<Array<ByteArray>>): ByteArray {
        val sigArr = legs.map { leg -> Cbor.A(listOf(Cbor.B(leg[0]), Cbor.M(listOf()), Cbor.B(leg[1]))) }
        return Cbor.encode(
            Cbor.Tag(
                TAG_SIGN,
                Cbor.A(listOf(Cbor.B(bodyProt), Cbor.M(listOf()), Cbor.B(payload), Cbor.A(sigArr)))
            )
        )
    }

    /** Recover the body protected header, payload, and ordered legs from a tagged COSE_Sign object. */
    fun parseSignRaw(obj: ByteArray): RotationParse {
        val v = Cbor.decode(obj)
        if (v !is Cbor.Tag || v.n != TAG_SIGN || v.content !is Cbor.A) {
            throw NaalpException("Malformed", "not a tagged COSE_Sign")
        }
        val items = v.content.items
        if (items.size != 4 || items[0] !is Cbor.B || items[2] !is Cbor.B || items[3] !is Cbor.A) {
            throw NaalpException("Malformed", "malformed COSE_Sign array")
        }
        val legs = ArrayList<Array<ByteArray>>()
        for (sv in (items[3] as Cbor.A).items) {
            if (sv !is Cbor.A || sv.items.size != 3 || sv.items[0] !is Cbor.B || sv.items[2] !is Cbor.B) {
                throw NaalpException("Malformed", "malformed COSE_Signature leg")
            }
            legs.add(arrayOf((sv.items[0] as Cbor.B).v, (sv.items[2] as Cbor.B).v))
        }
        return RotationParse((items[0] as Cbor.B).v, (items[2] as Cbor.B).v, legs)
    }

    /** Extract the alg (label 1) value from a serialized leg protected header {1: alg}. */
    fun algFromProtected(prot: ByteArray): Int {
        // §3.1.1 (R5): reject the redundant 0x41A0 encoding of an empty protected header (a bstr
        // wrapping an empty map; its unwrapped content is the single byte 0xA0) as NonCanonical,
        // before interpreting the header — the empty protected header is pinned to 0x40.
        if (prot.size == 1 && (prot[0].toInt() and 0xFF) == 0xA0) {
            throw NaalpException("NonCanonical", "empty protected header must be 0x40, not 0x41A0 (§3.1.1, R5)")
        }
        val v = Cbor.decode(prot)
        if (v !is Cbor.M) throw NaalpException("Malformed", "protected header not a map")
        for (p in v.pairs) {
            val k = p.k
            if (k is Cbor.U && k.v == 1L && p.v is Cbor.N) return (p.v as Cbor.N).v.toInt()
        }
        throw NaalpException("Malformed", "no alg in protected header")
    }

    // --- ML-DSA (FIPS 204) ---

    private fun mldsaParams(alg: Int): MLDSAParameters = when (alg) {
        ALG_MLDSA65 -> MLDSAParameters.ml_dsa_65
        ALG_MLDSA87 -> MLDSAParameters.ml_dsa_87
        else -> throw NaalpException("UnknownAlg", "alg $alg is not an ML-DSA algorithm")
    }

    /** Derive the public key from a 32-byte seed (NIST ACVP keyGen); returns pk bytes. */
    fun mldsaKeygen(param: String, seed: ByteArray): ByteArray {
        val p = if (param == "ML-DSA-87") MLDSAParameters.ml_dsa_87 else MLDSAParameters.ml_dsa_65
        if (seed.size != 32) throw NaalpException("Malformed", "ML-DSA seed must be 32 bytes")
        val sk = MLDSAPrivateKeyParameters(p, seed)
        return sk.publicKey
    }

    /** Deterministic (rnd=0) ML-DSA signature over tbs with the key derived from seed. */
    fun mldsaSign(alg: Int, seed: ByteArray, tbs: ByteArray): ByteArray {
        val p = mldsaParams(alg)
        if (seed.size != 32) throw NaalpException("Malformed", "ML-DSA seed must be 32 bytes")
        val sk = MLDSAPrivateKeyParameters(p, seed)
        val signer = MLDSASigner()
        signer.init(true, sk) // no ParametersWithRandom -> rnd = 32 zero bytes (deterministic)
        signer.update(tbs, 0, tbs.size)
        return try {
            signer.generateSignature()
        } catch (e: Exception) {
            throw NaalpException("SignFailed", e.toString())
        }
    }

    fun mldsaVerify(alg: Int, pk: ByteArray, tbs: ByteArray, sig: ByteArray): Boolean {
        val p = mldsaParams(alg)
        val pub = MLDSAPublicKeyParameters(p, pk)
        val signer = MLDSASigner()
        signer.init(false, pub)
        signer.update(tbs, 0, tbs.size)
        return signer.verifySignature(sig)
    }

    // --- Ed25519 (RFC 8032) ---

    fun ed25519Sign(seed: ByteArray, msg: ByteArray): ByteArray {
        if (seed.size != 32) throw NaalpException("Malformed", "ed25519 secret key must be a 32-byte seed")
        val priv = Ed25519PrivateKeyParameters(seed, 0)
        val signer = Ed25519Signer()
        signer.init(true, priv)
        signer.update(msg, 0, msg.size)
        return signer.generateSignature()
    }

    fun ed25519Verify(pk: ByteArray, msg: ByteArray, sig: ByteArray): Boolean {
        if (pk.size != 32) return false
        val pub = Ed25519PublicKeyParameters(pk, 0)
        val signer = Ed25519Signer()
        signer.init(false, pub)
        signer.update(msg, 0, msg.size)
        return signer.verifySignature(sig)
    }

    // --- LAMPS opt-in composite signature (alg -65537, design.md §4.2) ---

    const val ALG_COMPOSITE_65_ED25519 = -65537 // COMPSIG-MLDSA65-Ed25519-SHA512
    const val ALG_COMPOSITE_44_ED25519 = -65538 // edge; RESERVED, not implemented
    private val COMPOSITE_PREFIX = "CompositeAlgorithmSignatures2025".toByteArray(Charsets.US_ASCII)
    private val COMPOSITE_LABEL_MLDSA65_ED25519 = "COMPSIG-MLDSA65-Ed25519-SHA512".toByteArray(Charsets.US_ASCII)
    private const val MLDSA65_SIG_SIZE = 3309 // FIPS 204 ML-DSA-65 signature size
    const val MLDSA65_PUB_SIZE = 1952 // FIPS 204 ML-DSA-65 pubkey size (composite split point)

    /**
     * The LAMPS composite message representative M' = Prefix || Label || len(ctx) || ctx ||
     * SHA-512(M) (design.md §4.2). len(ctx) is a single length octet; the N-AALP composite context
     * is empty, so the octet is 0x00. Both legs sign this same M'.
     */
    fun computeMprime(label: ByteArray, ctx: ByteArray, m: ByteArray): ByteArray {
        if (ctx.size > 255) throw NaalpException("Malformed", "composite context exceeds one length octet")
        val h = MessageDigest.getInstance("SHA-512").digest(m)
        val out = ByteArray(COMPOSITE_PREFIX.size + label.size + 1 + ctx.size + h.size)
        var o = 0
        COMPOSITE_PREFIX.copyInto(out, o); o += COMPOSITE_PREFIX.size
        label.copyInto(out, o); o += label.size
        out[o++] = ctx.size.toByte() // len(ctx) as a single length octet
        ctx.copyInto(out, o); o += ctx.size
        h.copyInto(out, o)
        return out
    }

    /**
     * The LAMPS composite signature value over the COSE ToBeSigned tbs: mldsaSig || tradSig
     * (ML-DSA-65 first, raw concatenation; §4.2). The ML-DSA leg is deterministic (no
     * ParametersWithRandom => rnd=0) with context = the suite Label octets (via ParametersWithContext);
     * the Ed25519 leg signs M' with no context.
     */
    fun compositeSign(mldsaSeed: ByteArray, edSeed: ByteArray, tbs: ByteArray): ByteArray {
        if (mldsaSeed.size != 32) throw NaalpException("Malformed", "ML-DSA seed must be 32 bytes")
        val mprime = computeMprime(COMPOSITE_LABEL_MLDSA65_ED25519, ByteArray(0), tbs)
        val sk = MLDSAPrivateKeyParameters(MLDSAParameters.ml_dsa_65, mldsaSeed)
        val signer = MLDSASigner()
        signer.init(true, ParametersWithContext(sk, COMPOSITE_LABEL_MLDSA65_ED25519))
        signer.update(mprime, 0, mprime.size)
        val mldsaSig = try {
            signer.generateSignature()
        } catch (e: Exception) {
            throw NaalpException("SignFailed", e.toString())
        }
        val tradSig = ed25519Sign(edSeed, mprime)
        return mldsaSig + tradSig // ML-DSA first (LAMPS order)
    }

    /**
     * Valid IFF BOTH the ML-DSA-65 leg (context = Label) and the Ed25519 leg (no context) validate
     * over M'. A value of the wrong length is malformed and rejected. A stripped or re-interpreted
     * lone leg has no valid composite because M' binds both components (RFC 9955; §4.2/§4.5).
     */
    fun compositeVerify(mldsaPk: ByteArray, edPk: ByteArray, m: ByteArray, sig: ByteArray): Boolean {
        if (sig.size != MLDSA65_SIG_SIZE + 64) return false
        val mprime = computeMprime(COMPOSITE_LABEL_MLDSA65_ED25519, ByteArray(0), m)
        val pub = MLDSAPublicKeyParameters(MLDSAParameters.ml_dsa_65, mldsaPk)
        val signer = MLDSASigner()
        signer.init(false, ParametersWithContext(pub, COMPOSITE_LABEL_MLDSA65_ED25519))
        signer.update(mprime, 0, mprime.size)
        val mldsaOk = signer.verifySignature(sig.copyOfRange(0, MLDSA65_SIG_SIZE))
        val edOk = ed25519Verify(edPk, mprime, sig.copyOfRange(MLDSA65_SIG_SIZE, sig.size))
        return mldsaOk && edOk
    }

    /** Produce a deterministic tagged COSE_Sign1 object over (protected, payload). */
    fun coseSign1(alg: Int, seed: ByteArray, protectedHeader: ByteArray, payload: ByteArray): ByteArray {
        val tbs = toBeSignedRaw(protectedHeader, payload)
        val sig = mldsaSign(alg, seed, tbs)
        return assembleSign1Raw(protectedHeader, payload, sig)
    }

    /** Verify a raw signature over already-assembled ToBeSigned bytes, dispatching by alg. */
    fun coseVerify1Raw(alg: Int, pk: ByteArray, tbs: ByteArray, sig: ByteArray): Boolean = when (alg) {
        ALG_MLDSA65, ALG_MLDSA87 -> mldsaVerify(alg, pk, tbs, sig)
        ALG_ED25519 -> ed25519Verify(pk, tbs, sig)
        else -> throw NaalpException("UnknownAlg", "unknown alg $alg")
    }

    fun coseVerify1(alg: Int, pk: ByteArray, obj: ByteArray): Boolean {
        val parts = parseSign1Raw(obj)
        val tbs = toBeSignedRaw(parts[0], parts[1])
        return coseVerify1Raw(alg, pk, tbs, parts[2])
    }
}
